### PR TITLE
Type arguments should not be inferred to be nulltype

### LINKED ISSUE #887: error: AsSuperVisitor: Unexpected combination: type: NULL superType: DECLARED since v2.1.0
I've gotten error: `error: AsSuperVisitor: Unexpected combination: type: NULL superType: DECLARED` from checker-framework when trying to upgrade from using v2.0.0 to using v2.1.1.
I can repro this issue in **v2.1.0** but note the stack trace is from v2.1.1.

Unfortunately, due to company policies, I can't share the code.
And I'm not able to create a repro case because the stack trace doesn't help me much. Here are the error stack traces:

```
error: AsSuperVisitor: Unexpected combination: type: NULL superType: DECLARED.
  type:  null
  superType: ViewModel
  Compilation unit: java/com/some_package/SomeClass.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:739)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:28)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:39)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitNull_Declared(AbstractAtmComboVisitor.java:349)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:497)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:45)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:77)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:63)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:128)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:434)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:338)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:934)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:389)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:469)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:338)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:934)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:389)
  org.checkerframework.framework.util.QualifierPolymorphism.annotate(QualifierPolymorphism.java:204)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1200)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory.methodFromUse(KeyForAnnotatedTypeFactory.java:213)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:207)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:50)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:37)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1211)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:982)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:210)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:915)
  org.checkerframework.checker.nullness.KeyForTransfer.visitMethodInvocation(KeyForTransfer.java:56)
  org.checkerframework.checker.nullness.KeyForTransfer.visitMethodInvocation(KeyForTransfer.java:27)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:80)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:404)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:222)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:961)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:918)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:845)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1152)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:205)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:286)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:169)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:973)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:464)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:458)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:208)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
error: AsSuperVisitor: Unexpected combination: type: NULL superType: DECLARED.
  type: null
  superType: ViewModel
  Compilation unit: java/com/some_package/SomeClass.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:739)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:28)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:39)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitNull_Declared(AbstractAtmComboVisitor.java:349)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:497)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:45)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:77)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:63)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:128)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:434)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:338)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:934)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:389)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:469)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visitDeclared(QualifierPolymorphism.java:338)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:934)
  org.checkerframework.framework.type.visitor.SimpleAnnotatedTypeVisitor.visit(SimpleAnnotatedTypeVisitor.java:64)
  org.checkerframework.framework.util.QualifierPolymorphism$PolyCollector.visit(QualifierPolymorphism.java:389)
  org.checkerframework.framework.util.QualifierPolymorphism.annotate(QualifierPolymorphism.java:204)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1200)
  org.checkerframework.checker.nullness.NullnessAnnotatedTypeFactory.methodFromUse(NullnessAnnotatedTypeFactory.java:308)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:207)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:50)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:37)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1211)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:982)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:210)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:915)
  org.checkerframework.checker.initialization.InitializationTransfer.visitMethodInvocation(InitializationTransfer.java:198)
  org.checkerframework.checker.nullness.NullnessTransfer.visitMethodInvocation(NullnessTransfer.java:214)
  org.checkerframework.checker.nullness.NullnessTransfer.visitMethodInvocation(NullnessTransfer.java:56)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:80)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:404)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:222)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:961)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:918)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:845)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1152)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:205)
  org.checkerframework.checker.initialization.InitializationVisitor.visitClass(InitializationVisitor.java:276)
  org.checkerframework.checker.initialization.InitializationVisitor.visitClass(InitializationVisitor.java:59)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:973)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:464)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:208)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
```

### PR DESCRIPTION
**Pull request by smillst**
If an argument to a method is null, then the JLS says that it does not constrain the type argument.  However, null may constrain the qualifiers on the type argument, so it is included in the constraints but is not used as the underlying type of the type argument.

Fixes #887  (tested with daikon-typecheck)

### CODE REVIEW

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 1]
```diff
@@ -635,8 +640,13 @@ private void handleUninferredTypeVariables(
                 final AnnotatedTypeMirror inferredType = inferredArgs.get(typeVar);
 
                 if (inferredType == null) {
-                    AnnotatedTypeMirror dummy = typeFactory.getUninferredWildcardType(atv);
-                    inferredArgs.put(atv.getUnderlyingType(), dummy);
+                    AnnotatedTypeMirror dummy = atv.getUpperBound().deepCopy();
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: The upper bound of the type variable might refer to other type variables. It seems wrong to directly use the upper bound as inferred argument. Why is this correct?

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 2]
```diff
@@ -635,8 +640,13 @@ private void handleUninferredTypeVariables(
...
+                    inferredArgs.put(typeVar, dummy);
+                } else if (inferredType.getKind() == TypeKind.NULL) {
+                    // Type argument cannot be null type, use upper bound instead.
+                    AnnotatedTypeMirror dummy = atv.getUpperBound().deepCopy();
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Same comment as above.

File: framework/tests/all-systems/InferNullType.java [hunk 3]
```diff
@@ -4,8 +4,13 @@
         return input;
     }
 
+    <T> T toInfer2(T input) {
```
Comment (wmdietl) [inline framework/tests/all-systems/InferNullType.java]: Please add examples with multiple type variables, e.g `<T extends @Nullable Number, S extends T>` and then try null in the different possible combinations.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java (hunk 1 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: I don't understand what seems wrong about using the upper bound.  Previously, we created a new wildcard and set its upper bound to the upper bound of the type var.  The upper bound of that type variable might also have been a type variable.

File: framework/tests/all-systems/InferNullType.java (hunk 3 above)
Comment (smillst) [inline framework/tests/all-systems/InferNullType.java]: These test cases caused a crash, which I fixed.  The test cases and fix have been committed.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 4]
```diff
@@ -370,7 +425,8 @@ private void clampToLowerBound(
             final AnnotatedTypeFactory typeFactory,
             final List<AnnotatedTypeMirror> argTypes,
             final AnnotatedExecutableType methodType,
-            final Set<TypeVariable> targets) {
+            final Set<TypeVariable> targets,
+            boolean useNullArguments) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Some documentation would be good for the new parameter. checker-framework-inference is using this method and fails to compile with this change.

**Pull request merged**